### ISSUE TITLE
EntityRemoved called during clientside full updates

### ISSUE THREAD

**Issue by WilliamVenner** [Facepunch/garrysmod-issues#4675]
**Problem**

So whilst I acknowledge this is "expected behaviour", there are a few problems with this:

1. What should be a [simple hook](https://wiki.facepunch.com/gmod/GM:EntityRemoved) has been undocumented to this complexity for (probably) years (with the exception of [ENTITY:OnRemove](https://wiki.facepunch.com/gmod/ENTITY:OnRemove)) so there are many scripts which use this hook which could break after a full update
2. It gets worse when you consider that [Entity:CallOnRemove](https://wiki.facepunch.com/gmod/Entity:CallOnRemove) is also [vulnerable to this issue](https://github.com/Facepunch/garrysmod/blob/bc5b70e8e0ead1b32b2b29809f8ce15aa7b0888c/garrysmod/lua/includes/extensions/entity.lua#L142) and there are likely many scripts that overlook this documented behaviour for [ENTITY:OnRemove](https://wiki.facepunch.com/gmod/ENTITY:OnRemove)
3. Knowing when a full update is happening is an extremely niche edge case and doesn't match the assumed behaviour of these entity removed hooks, which should be as simple as they are named.

**Solution**

It would make much more sense if this hook (and related hooks) were called simply in the way they are named, and if it's really needed a second hook is created to move the native source engine behaviour of this hook into a more technical hook which is properly documented.

And of course, other quirks with these hooks should be warned & documented (e.g. client PVS-related fuckery)

If it's just not possible, or technically infeasible, then the documentation for these hooks really need to be updated clearly.

**Evidence of scripts that could be vulnerable to this "issue":**

- https://github.com/search?l=Lua&q=hook+entityremoved&type=Code
- https://github.com/search?l=Lua&q=callonremove&type=Code

**Comment by thegrb93** [comment]
Related https://github.com/Facepunch/garrysmod-issues/issues/2637

**Comment by WilliamVenner** [comment]
Workaround for anyone stumbling on this issue

```lua
function ENT:OnRemoved()
    timer.Simple(0, function()
        if not IsValid(self) then
            -- We actually got removed
        else
            -- It was just a full update
        end
    end)
end

Entity:CallOnRemove("Example", function(ent)
    timer.Simple(0, function()
        if not IsValid(ent) then
            -- We actually got removed
        else
            -- It was just a full update
        end
    end)
end)

hook.Add("OnRemove", "Example", function(ent)
    timer.Simple(0, function()
        if not IsValid(ent) then
            -- We actually got removed
        else
            -- It was just a full update
        end
    end)
end)
```

**Comment by RaphaelIT7** [comment]
Currently, `OnRemove` and `EntityRemoved` won't be called while the Entities still become NULL.
Example: Players outside the PVS and all their children become NULL until the next player forced update(Without calling `OnRemove` & `EntityRemoved`) which broke a few things for me and maybe for others too.

**Comment by robotboy655** [comment]
Are you reporting a regression, or a separate issue?

**Comment by RaphaelIT7** [comment]
It's somewhat both but more a regression because the bug fix caused Entities outside the PVS to become NULL without any Lua callbacks and only after the next forced update ~10 seconds(bound to sv_playerforcedupdate) they are Valid again because players and their children are not networked on a full update.

**Comment by robotboy655** [comment]
That makes no sense, the fix did nothing that would cause entities to become NULL, it only make the hooks not be called on a full update.

**Comment by RaphaelIT7** [comment]
The full update causes them to become NULL, but the bug fix removed most ways to check it on a full update.
```lua
hook.Add("NetworkEntityCreated", "Example", function(ent)
	if GetClass(ent) == "weapon_example" then
		table.insert(weps, ent)
	end
end)

hook.Add("EntityRemoved", "Example", function(ent) -- Not called anymore and the entity won't be removed
	if GetClass(ent) == "weapon_example" then
		table.RemoveByValue(weps, ent)
	end
end)
```
It would maybe be better if `NetworkEntityCreated`,  `EntityRemoved` and `Entity:OnRemove` just get an additional argument full update(true or false) which would allows one to simply check if it's a full update, and it would remove the need for any useless workaround that are currently needed or at least if the entity is outside the PVS/dormant call EntityRemoved because it could be that it won't be included in the full update and because of this be NULL until it re-enters the PVS

**Comment by thegrb93** [comment]
> It would maybe be better if `NetworkEntityCreated`,  `EntityRemoved` and `Entity:OnRemove` just get an additional argument full update(true or false) which would allows one to simply check if it's a full update

No that's a stupid workaround

**Comment by Kefta** [comment]
That sounds like a pretty good workaround to continue supporting the many pre-fullupdate hook solutions.

**Comment by thegrb93** [comment]
Just don't let entities become NULL from a full update.

>That makes no sense, the fix did nothing that would cause entities to become NULL, it only make the hooks not be called on a full update.

This always happened, it isn't related to your fix. It's just another weird thing that also needs fixing.

**Comment by Person8880** [comment]
The changes to `ENT:OnRemove` seem to have caused a regression in the case where:

1. An entity leaves the player's PVS.
2. A full update is received.

In this case, it seems logical that `ENT:OnRemove` would be called, as the entity is destroyed on the client and may never be seen again if the client never re-enters the PVS within which it exists. However, with the changes made this is no longer the case, and thus any resources associated with the entity that a script expects to clean up in `ENT:OnRemove` are never dealt with. 

The `EntityRemoved` hook is triggered, with `true` for the second argument, so in theory the old behaviour can be restored through this hook. However, that then also triggers for entities that are destroyed and re-created by the full update that have not left the PVS, taking us back to square one.

As an aside, `NetworkEntityCreated` is still called for entities that remain within the PVS after a full update. This was used previously as a mechanism to detect such entities and restore them. However, now that `ENT:OnRemove` is not called, this hook can trigger initialisation behaviour twice, which may also break scripts that don't expect it and don't have idempotent initialisation logic in their entities (e.g. creating a sound on the entity twice, with the old sound getting garbage collected and thus stopping playback after a short time).

**Comment by robotboy655** [comment]
My first instinct was also to have the Lua state be completely destroyed, but I saw code that is aimed to directly avoid that on full updates, which is why I went the other way.

I honestly don't see a good solution for this issue at this point. No matter what I do, there will be unwanted consequences.

I think moving forward the only option I see is having the extra argument for full update removals, and document this behavior on wiki better.
